helix-core/src/main/java/org/apache/helix/HelixManagerFactory.java

```diff
@@ -65,4 +64,39 @@ public static HelixManager getZKHelixManager(String clusterName, String instance
     return new ZKHelixManager(clusterName, instanceName, type, zkAddr, stateListener);
   }
 
+  /**
+   * Construct a ZkHelixManager using the HelixManagerProperty instance given.
+   * HelixManagerProperty given must contain a valid ZkConnectionConfig.
+   * @param clusterName
+   * @param instanceName
+   * @param type
+   * @param stateListener
+   * @param helixManagerProperty must contain a valid ZkConnectionConfig
+   * @return
+   */
+  public static HelixManager getZKHelixManager(String clusterName, String instanceName,
+      InstanceType type, HelixManagerStateListener stateListener,
+      HelixManagerProperty helixManagerProperty) {
+    return new ZKHelixManager(clusterName, instanceName, type, null, stateListener,
+        helixManagerProperty);
+  }
+
+  /**
+   * Construct a ZkHelixManager using the HelixManagerProperty instance given.
+   * NOTE: if both zkAddr and a valid ZkConnectionConfig are given in HelixManagerProperty, the
+   * instantiation will fail - only one is required.
+   * @param clusterName
+   * @param instanceName
+   * @param type
+   * @param zkAddr
+   * @param stateListener
+   * @param helixManagerProperty
+   * @return
+   */
+  public static HelixManager getZKHelixManager(String clusterName, String instanceName,
+      InstanceType type, String zkAddr, HelixManagerStateListener stateListener,
+      HelixManagerProperty helixManagerProperty) {
```

Comment:
Can we get rid of this constructor by allowing setup zkAddr in the HelixManagerProperty or in the RealmAwareZkConnectionConfig?

Reply:
In addition, if we do so, then the HelixPropertyFactory can call the property builder to build a property even for the existing getHelixProperty API, right?
So it would be clean,
1. Callers only call factory methods to get property.
2. The factory calls the property builder to get the new object.

Reply:
RealmAware zk config should not have a field called ZkAddress. That's against the design of RealmAware API.

Reply:
As to whether we should "move" `zkAddress` into helixManagerProperty - this idea is not ideal. It is what we have now and we should keep it in the constructor for backward-compatibility, and giving users multiple ways to input `zkAddress` will undoubtedly confuse users. 

So I think the right thing to do here is to keep zkAddress in HelixManager/HelixManagerFactory as a parameter. It won't continue to increase getZkHelixManager() methods. All future configs that are needed to generate an instance of HelixManager could go into `helixManagerProperty`.

Reply:
I think we shall not be blocked on this zkAddress argument. It can be discussed later.

How about this, to ensure we are not introducing more confusing, we don't this specific method. So for this class, either you give a zkAddress to get HM or you give a HelixManagerProperty to get the HM.

Reply:
Have you considered: what if a user needs to provide both `zkAddr` and `helixManagerProperty`?

At any rate, we could add this constructor later when we absolutely need to. For the sake of keeping this moving forward, I'll remove this `getZkHelixManager` method for now.

Reply:
My thought is the same as what you described. We can always add it later if needed.

Reply:
It seems this comment has been marked as resolved without finishing the code change. Are you still working on this change?